### PR TITLE
[ParameterUpdate] Part 1: ParameterAggregate protocol and synthesis.

### PR DESCRIPTION
**Pull request by dan-zheng**
The `ParameterAggregate` protocol represents an aggregate of parameters.
Types that conform to `ParameterAggregate` must specify a `Parameter`
associated type and an update method.

For nominal types whose stored properties all have the same type,
the compiler can synthesize the `ParameterAggregate` protocol requirements.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -2381,6 +2381,9 @@ ERROR(broken_encodable_requirement,none,
       "Encodable protocol is broken: unexpected requirement", ())
 ERROR(broken_decodable_requirement,none,
       "Decodable protocol is broken: unexpected requirement", ())
+// SWIFT_ENABLE_TENSORFLOW
+ERROR(broken_parameter_aggregate_requirement,none,
+      "ParameterAggregate protocol is broken: unexpected requirement", ())
```
Comment (rxwei) [inline include/swift/AST/DiagnosticsSema.def]: What do you mean by "broken"? This diagnostic is not understandable by the user.

File: lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -1739,7 +1740,9 @@ ASTContext::getModule(ArrayRef<std::pair<Identifier, SourceLoc>> ModulePath) {
     if (ModuleDecl *M = importer->loadModule(moduleID.second, ModulePath)) {
       if (ModulePath.size() == 1 &&
           (ModulePath[0].first == StdlibModuleName ||
-           ModulePath[0].first == Id_Foundation))
+           ModulePath[0].first == Id_Foundation ||
+           // SWIFT_ENABLE_TENSORFLOW
+           ModulePath[0].first == getIdentifier("TensorFlow")))
```
Comment (rxwei) [inline lib/AST/ASTContext.cpp]: Add `ID_TensorFlow`?

File: include/swift/Serialization/ModuleFormat.h [hunk 3]
```diff
@@ -55,7 +55,7 @@ const uint16_t VERSION_MAJOR = 0;
 /// describe what change you made. The content of this comment isn't important;
 /// it just ensures a conflict if two people change the module format.
 /// Don't worry about adhering to the 80-column limit for this line.
-const uint16_t VERSION_MINOR = 427; // SWIFT_ENABLE_TENSORFLOW: serialize @differentiable.
+const uint16_t VERSION_MINOR = 428; // SWIFT_ENABLE_TENSORFLOW: ParameterAggregate.
```
Comment (rxwei) [inline include/swift/Serialization/ModuleFormat.h]: I don't think anything changed the module format.

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/DiagnosticsSema.def]: Removed diagnostic and replaced usage with assertions in https://github.com/apple/swift/pull/18140/commits/7768fa9d6d142cf111f4f110d44b68dd1cf52a74.
This is an internal error, I originally copied the diagnostic from other synthesis-related diagnostics.

File: lib/AST/ASTContext.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/AST/ASTContext.cpp]: Done. There are many usages of `getIdentifier("TensorFlow")`, I'll replace them in a follow-up.

File: include/swift/Serialization/ModuleFormat.h (hunk 3 above)
Comment (dan-zheng) [reply include/swift/Serialization/ModuleFormat.h]: Reverted in https://github.com/apple/swift/pull/18140/commits/7768fa9d6d142cf111f4f110d44b68dd1cf52a74.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (rxwei) [reply include/swift/AST/DiagnosticsSema.def]: I see! I missed those precedents. Either way is fine I think.

**Comment by rxwei** [comment]
Great work! Now that we have the implementation, it's time to decide whether `@parameter` is a good name. "Parameter" is going to sound like a lot of other things in normal programming. 

I'd suggest `@TFParameter`. It doesn't look very native to Swift, but it's consistent with precedents such as `@IBOutlet`.

**Pull request merged**